From 87a1623fb268abb6696448f6132334e941025c22 Mon Sep 17 00:00:00 2001
From: DooMLoRD <metoo.mememe@gmail.com>
Date: Sun, 18 May 2014 02:53:48 +0530
Subject: [PATCH] intelli-thermal: use its own dedicated workqueue

    this should reduce the load on the shared global workqueue for other
    processes and also give its own unique variable name for cpu limit

    Signed-off-by: Paul Reioux <reioux@gmail.com>
---
 drivers/thermal/intelli/msm_thermal.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/drivers/thermal/intelli/msm_thermal.c b/drivers/thermal/intelli/msm_thermal.c
index 648d9aa..b1a34a5 100644
--- a/drivers/thermal/intelli/msm_thermal.c
+++ b/drivers/thermal/intelli/msm_thermal.c
@@ -43,6 +43,7 @@ static int enabled;
 static struct msm_thermal_data msm_thermal_info;
 static uint32_t limited_max_freq_thermal = MSM_CPUFREQ_NO_LIMIT;
 static struct delayed_work check_temp_work;
+static struct workqueue_struct *intellithermal_wq;
 static bool core_control_enabled;
 static uint32_t cpus_offlined;
 static DEFINE_MUTEX(core_control_mutex);
@@ -252,7 +253,7 @@ static void __ref check_temp(struct work_struct *work)
 	//pr_info("msm_thermal: worker is alive!\n");
 reschedule:
 	if (enabled)
-		schedule_delayed_work(&check_temp_work,
+		queue_delayed_work(intellithermal_wq, &check_temp_work,
 				msecs_to_jiffies(msm_thermal_info.poll_ms));
 }
 
@@ -289,10 +290,7 @@ static void __ref disable_msm_thermal(void)
 {
 	int cpu = 0;
 
-	
-	cancel_delayed_work_sync(&check_temp_work);
-	//flush_scheduled_work();
-
+	flush_workqueue(intellithermal_wq);
 
 	for_each_possible_cpu(cpu) {
 		update_cpu_max_freq(cpu, MSM_CPUFREQ_NO_LIMIT);
@@ -310,7 +308,8 @@ static int __ref set_enabled(const char *val, const struct kernel_param *kp)
 	} else {
 		if (!enabled) {
 			enabled = 1;
-			schedule_delayed_work(&check_temp_work, 0);
+			queue_delayed_work(intellithermal_wq,
+					   &check_temp_work, 0);
 			pr_info("msm_thermal: rescheduling...\n");
 		} else
 			pr_info("msm_thermal: already running...\n");
@@ -578,8 +577,10 @@ int __init msm_thermal_init(struct msm_thermal_data *pdata)
 	enabled = 1;
 	if (num_possible_cpus() > 1)
 		core_control_enabled = 1;
+	intellithermal_wq = alloc_workqueue("intellithermal",
+				WQ_UNBOUND | WQ_MEM_RECLAIM, 1);
 	INIT_DELAYED_WORK(&check_temp_work, check_temp);
-	schedule_delayed_work(&check_temp_work, 0);
+	queue_delayed_work(intellithermal_wq, &check_temp_work, 0);
 
 	if (num_possible_cpus() > 1)
 		register_cpu_notifier(&msm_thermal_cpu_notifier);
